rspace/src/main/scala/coop/rchain/rspace/StableHashProvider.scala

```diff
@@ -0,0 +1,42 @@
+package coop.rchain.rspace
+
+import scala.collection.immutable.Seq
+import cats.implicits._
+import scodec.{Attempt, Codec}
+import scodec.bits.BitVector
+import scodec.codecs._
+import scodec.interop.cats._
+import coop.rchain.shared.AttemptOps._
+
+object StableHashProvider {
+  def hash[C](channels: Seq[C])(implicit codecC: Codec[C]): Blake2b256Hash =
+    Blake2b256Hash.create(
+      channels
+        .map(c => codecC.encode(c).get.bytes)
+        .sorted(util.ordByteVector))
+
+  def hash[C, P, K](channels: Seq[C], patterns: Seq[P], continuation: K, persist: Boolean)(
+      implicit
+      serializeC: Serialize[C],
+      serializeP: Serialize[P],
+      serializeK: Serialize[K]) = {
+    val (channelsSeq, patternsSeq) = channels
+      .zip(patterns)
+      .map { case (channel, pattern) => (serializeC.encode(channel), serializeP.encode(pattern)) }
+      .sorted(util.ordByteVectorPair)
+      .unzip
+
+    Blake2b256Hash.create(
+      channelsSeq ++ patternsSeq
+        :+ serializeK.encode(continuation)
```

Comment:
@birchmd pointed out to me that the default `immutable.Seq` is a `List`, so these single-element appends are expensive.

probably better to put them into a two element `Seq` and append that to the others.